### PR TITLE
Do while task support in builder

### PR DESCRIPTION
**Pull request by marosmars**
Do while task in conductor:
https://netflix.github.io/conductor/configuration/systask/#do-while-task

It is a simple while loop.
It cannot have nested loops or subworkflows in it.

Do while is not present in FM version of conductor and is disabled by
default.

Signed-off-by: Maros Marsalek <mmarsalek@frinx.io>

Screenshots:

dowhile_1
dowhile_2
dowhile_3
dowhile_4
dowhile_5

### CODE REVIEW

**Comment by marosmars** [comment]
@ChmaraX I think I've modified all the places required for a new built-in task.
I also tested it manually with a bunch of workflows designed by hand.

but pls review if you can

**Comment by marosmars** [comment]
Also, it wasn't easy adding support for a new built-in task ... I was thinking if we could do/think about refactoring the codebase to make it easier.

Currently the parsing (or model - model translation) is spread in too many places.
There are actually 3 different models for the workflow: UI model for react-diagrams, WFE model for workflow layout and the Conductor JSON model. But only one of them is formally defined (as classes and types) and that's the UI model for react-diagrams. Maybe we should do the same for the other models and then refactor the translations and put them under a single place (and use some patterns /if there are any/ for model2model translation).

and also add unit tests before doing anything, since without unit tests it is impossible to do big refactoring ...

what do you think ?

**Comment by ChmaraX** [comment]
> what do you think ?

I agree and I'm well aware of what should be refactored (and its sadly not just this) but since this app was initially made in few weeks as a demo without prior specification (or any software dev method at all) and since then, only new features were added by multiple people (also without specification and testing), it now requires some design patterns and greater effort to improve code quality which im trying think of and slowly apply now on the go, since I was not familiar (and probably noone involved) with Javascript or React patterns and best practices at the time of writing this project and still I would be glad to receive feedback/suggestions from more senior members on actual code implementation rather than on just visual side :)

TL'DR: it's a good idea, it should be discussed and prioritized, but it should come as a specific task, it doesn't matter what I think since its not my personal project

**Comment by marosmars** [comment]
I created an issue for refactoring
https://github.com/FRINXio/frinx-workflow-ui/issues/71

We should list specific things that need to be refactored/fixed ... then we can decide what to do with it 
So if you have anything specific, please list it there

File: src/pages/diagramBuilder/WorkflowDiagram.js [hunk 1]
```diff
@@ -839,6 +896,124 @@ export class WorkflowDiagram {
...
+          link = this.getLoopedNodeRightLink(parentNode)
+
+          // Process all tasks in loop until while_end
+          let isWhileEnd = () => loopTasks.length >= 1
+            && loopTasks[loopTasks.length - 1]
+            && loopTasks[loopTasks.length - 1].type === "DO_WHILE_END";
+
+          while (!isWhileEnd()) {
+            // Recursively parse nested task in loop
+            parentNode = this.parseTaskToJSON(link, parentNode, loopTasks);
```
Comment (ChmaraX) [inline src/pages/diagramBuilder/WorkflowDiagram.js:963]: `link` is `null` in probably last iteration, therefore it cant finish traversing through nodes and finish mapping of nodes into JSON

`Error: Cannot read property 'sourcePort' of null`

**Review by ChmaraX — CHANGES_REQUESTED** [review]
I think it would be maybe better to have loop condition in "Input Parameters" tab or outside "Advanced parameters" similiar to other system tasks.

File: src/pages/diagramBuilder/WorkflowDiagram.js (hunk 1 above)
Comment (marosmars) [reply src/pages/diagramBuilder/WorkflowDiagram.js:961]: Fixed finally, isWhileEnd was checking wrong task type ... do_while_end instead of while_end

**Comment by marosmars** [comment]
> I think it would be maybe better to have loop condition in "Input Parameters" tab or outside "Advanced parameters" similiar to other system tasks.

I guess ... but it is not an inputParameter, the task definition requires that outside of inputParams. So the fact that it ends up in advanced parameters is because it is not in inputParameters as defined in Sidebar.js

Would that still be possible ?

**Review by ChmaraX — APPROVED** [review]

**Pull request merged**